dlt/pipeline/pipeline.py

```diff
@@ -1063,8 +1077,33 @@ def _wipe_working_folder(self) -> None:
     def _attach_pipeline(self) -> None:
         pass
 
+    def _refresh_source(self, source: DltSource) -> Tuple[Schema, TPipelineState, Dict[str, Any]]:
+        if self.refresh is None or self.first_run:
+            return source.schema, self.state, {}
+        _resources_to_drop = (
+            list(source.resources.extracted) if self.refresh != "drop_dataset" else []
+        )
+        drop_result = drop_resources(
+            source.schema,
+            self.state,
+            resources=_resources_to_drop,
+            drop_all=self.refresh == "drop_dataset",
+            state_paths="*" if self.refresh == "drop_dataset" else [],
+        )
+        load_package_state = {}
+        if drop_result.dropped_tables:
+            key = "dropped_tables" if self.refresh != "drop_data" else "truncated_tables"
+            load_package_state[key] = drop_result.dropped_tables
+        return drop_result.schema, drop_result.state, load_package_state
+
     def _extract_source(
-        self, extract: Extract, source: DltSource, max_parallel_items: int, workers: int
+        self,
+        extract: Extract,
+        source: DltSource,
+        max_parallel_items: int,
+        workers: int,
+        with_refresh: bool = False,
+        load_package_state_update: Optional[Dict[str, Any]] = None,
```

Comment:
@steinitzu @sh-rp IMO we should unify how we  generate package state in the pipeline. right now there's a hack we use to save pipeline state into package state at the end of extract
```
for source in data_to_sources(
                    data,
                    self,
                    schema,
                    table_name,
                    parent_table_name,
                    write_disposition,
                    columns,
                    primary_key,
                    schema_contract,
                ):
                    if source.exhausted:
                        raise SourceExhausted(source.name)

                    self._extract_source(
                        extract_step, source, max_parallel_items, workers, with_refresh=True
                    )
                # extract state
                state: TPipelineStateDoc = None
                if self.config.restore_from_destination:
                    # this will update state version hash so it will not be extracted again by with_state_sync
                    state = self._bump_version_and_extract_state(
                        self._container[StateInjectableContext].state, True, extract_step
                    )
                # commit load packages with state
                extract_step.commit_packages(state)
```
here:                 extract_step.commit_packages(state)

and there's a mechanism from @steinitzu to inject state during extract:
```
def extract(
....
      if load_package_state_update:
                    load_package.state.update(load_package_state_update)  # type: ignore[typeddict-item]
```
IMO we do not need both. and what steini did seems more generic so if we just add pipeline state to `load_package_state_update`
```
if should_extract and extract_state:
            data, doc = state_resource(state)
            extract_ = extract or Extract(
                self._schema_storage, self._normalize_storage_config(), original_data=data
            )
            self._extract_source(
                extract_,
                data_to_sources(data, self, schema or self.default_schema)[0],
                1,
                1,
                load_package_state_update=load_package_state_update,
```
under the right key, everything well work (and we fix one inconsistency - if we have many sources we add pipeline state to all packages, not only to the one that contains state document)
@sh-rp WDYT?

Reply:
yes, that sounds good to me. I was thinking we should put the pipeline state into the load package state always and let the destination figure out wether it needs to be stored based on the hash (right now it's only the filesystem that is using this anyway)

Reply:
Makes sense. Could we then not extract this `state_resource`? I.e. for drop command just create an empty load package with state and schema and no sources.

Reply:
@steinitzu sometimes we must update the state. right now AFAIK you somehow force the state hash to change so state is always extracted. you can try empty package (when there's not state change). I expect it will be ignored by the normalizer. but actually we should push it to load. you can try to fix it

Reply:
So I tried this, most of the changs in https://github.com/dlt-hub/dlt/pull/1063/commits/94020900aafc70b22712a91fc6b1e70d0c3fd15d  
State table is by default created like other `_dlt_` tables and state is inserted in destination in `complete_load()`  

Loading "empty" load package works (only load package state and new schema), but I had to remove this in normalize:
```python
            # if len(schema_files) == 0:
            #     # delete empty package
            #     self.normalize_storage.extracted_packages.delete_package(load_id)
            #     logger.info(f"Empty package {load_id} processed")
            #     continue
```
Is this just optimization or some other reason to skip empty packages? Would it be better to load it conditionally, e.g. with some marker file that marks it as "state only" package?

Reply:
@steinitzu I do not see any changes in `normalize` so let's keep it like that. we'll have a bigger push to reorganize how we write state